modules/codegen/src/main/scala/com/twilio/guardrail/generators/Scala/AkkaHttpServerGenerator.scala

```diff
@@ -166,42 +193,58 @@ object AkkaHttpServerGenerator {
         handlerName: String,
         methodSigs: List[scala.meta.Decl.Def],
         handlerDefinitions: List[scala.meta.Stat],
-        responseDefinitions: List[scala.meta.Defn]
+        responseDefinitions: List[scala.meta.Defn],
+        customExtraction: Boolean
     ) =
       for {
         _ <- Target.log.debug(s"renderHandler(${handlerName}, ${methodSigs}")
-      } yield q"""
+      } yield {
+        val tParams = if (customExtraction) List(customExtractionTypeParamContravariant) else List()
+        q"""
           trait ${Type.Name(handlerName)} {
             ..${methodSigs ++ handlerDefinitions}
           }
-        """
-    def getExtraRouteParams(tracing: Boolean) =
+        """.copy(tparams = tParams)
+      }
+    def getExtraRouteParams(customExtraction: Boolean, tracing: Boolean) =
       for {
         _ <- Target.log.debug(s"getExtraRouteParams(${tracing})")
-        res <- if (tracing) {
+        extractParam <- if (customExtraction) {
+          Target.pure(List(param"""customExtract: String => Directive1[$customExtractionTypeParamName]"""))
+        } else Target.pure(List.empty)
+        traceParam <- if (tracing) {
           Target.pure(List(param"""trace: String => Directive1[TraceBuilder]"""))
         } else Target.pure(List.empty)
-      } yield res
+      } yield extractParam ::: traceParam
     def renderClass(
         resourceName: String,
         handlerName: String,
         annotations: List[scala.meta.Mod.Annot],
         combinedRouteTerms: List[scala.meta.Stat],
         extraRouteParams: List[scala.meta.Term.Param],
         responseDefinitions: List[scala.meta.Defn],
-        supportDefinitions: List[scala.meta.Defn]
+        supportDefinitions: List[scala.meta.Defn],
+        customExtraction: Boolean
     ): Target[List[Defn]] =
       for {
         _ <- Target.log.debug(s"renderClass(${resourceName}, ${handlerName}, <combinedRouteTerms>, ${extraRouteParams})")
-        routesParams = List(param"handler: ${Type.Name(handlerName)}") ++ extraRouteParams
       } yield {
+        val handlerType =
+          if (customExtraction) {
+            Type.Apply(Type.Name(handlerName), List(customExtractionTypeParamName))
+          } else {
+            Type.Name(handlerName)
+          }
+        val typeParams     = if (customExtraction) customExtractionTypeParams else List()
+        val routesParams   = List(param"handler: $handlerType") ++ extraRouteParams
         val routeImplicits = List(param"implicit mat: akka.stream.Materializer") ++ AkkaHttpHelper.protocolImplicits(modelGeneratorType)
+        val routesDef      = q"""def routes(..${routesParams})(..$routeImplicits): Route = {
+                             ..${combinedRouteTerms}
+                           }""".copy(tparams = typeParams)
```

Comment:
oh dang, https://github.com/scalameta/scalameta/issues/519 got merged! Gotta go back and clean up these `copy(tparams =...`

Reply:
Rebased this change up

Reply:
Oh nice, I even found that PR while trying to figure this out, but didn't realize support had been merged :+1: .